langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java

```diff
@@ -342,13 +376,48 @@ private void createCollection(MongoDatabase database, String collectionName, Cre
         database.createCollection(collectionName, createCollectionOptions);
     }
 
-    private boolean isIndexExist(String indexName) {
+    private boolean indexExists(String indexName) {
+        Document indexRecord = indexRecord(collection, indexName);
+        return indexRecord != null && !indexRecord.getString("status").equals("DOES_NOT_EXIST");
+    }
+
+    private static Document indexRecord(
+            MongoCollection<MongoDbDocument> collection, String indexName) {
         return StreamSupport.stream(collection.listSearchIndexes().spliterator(), false)
-                .anyMatch(index -> indexName.equals(index.getString("name")));
+                .filter(index -> indexName.equals(index.getString("name")))
+                .findAny().orElse(null);
     }
 
     private void createIndex(String indexName, IndexMapping indexMapping) {
-        Document index = fromIndexMapping(indexMapping);
-        collection.createSearchIndex(indexName, index);
+        collection.createSearchIndexes(List.of(new SearchIndexModel(
+                indexName,
+                fromIndexMapping(indexMapping),
+                SearchIndexType.vectorSearch())));
+
+        waitForIndex(collection, indexName);
+    }
+
+    static void waitForIndex(MongoCollection<MongoDbDocument> collection, String indexName) {
+        long startTime = System.nanoTime();
+        long timeoutNanos = TimeUnit.SECONDS.toNanos(SECONDS_TO_WAIT_FOR_INDEX);
+        while (System.nanoTime() - startTime < timeoutNanos) {
+            Document indexRecord = indexRecord(collection, indexName);
+            if (indexRecord != null) {
+                if ("FAILED".equals(indexRecord.getString("status"))) {
+                    throw new RuntimeException("Search index has failed status.");
+                }
+                if (indexRecord.getBoolean("queryable")) {
+                    return;
+                }
+            }
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }
+        log.warn("Index {} was not created or did not exit INITIAL_SYNC within {} seconds",
+                indexName, SECONDS_TO_WAIT_FOR_INDEX);
```

Comment:
We wait on this condition. 
`index["type"] == "vectorSearch" and index["status"] == "READY"`

Determining when all of your new documents have been indexed is a more difficult challenge. The only thing I've found that works is to loop over vector searches and make sure than I get back the number that I requested, what you call cleverly, maxRequests.

Reply:
Still looking into this. If there is a difference, it seems like it will be subtle.

Reply:
This is fine. The only relevant difference is that a failed index might be queryable, but we check for failure directly above, so there is no difference here. Resolving.